File: guests/haskell/reorder.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DeriveAnyClass #-}

-- The reorder scene, Haskell port — guests/rust/reorder.rs,
-- tools/scenes/reorder.steps.

import GHC.Generics (Generic)

import Data.Text (Text)
import KayaApp

data Item = Item {title :: Text}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)


main :: IO ()
main = kayaMain $ \app -> do
  buildTx app $ do
    items <- collectionOf @Item

    let onRotate = submitTx app $ do
          entries <- recordItems items
          case entries of
            (firstKey, _) : _ -> moveToEnd (recordHandle items) firstKey
            [] -> return ()
        onLift = submitTx app $ do
          -- Keys, never indices.
          entries <- recordItems items
          case reverse entries of
            (lastKey, _) : _ -> moveToFront (recordHandle items) lastKey
            [] -> return ()

    root <-
      row
        [ buttonOn "rotate" onRotate,
          buttonOn "lift" onLift,
          each (recordHandle items) $ label (field @"title" @Item)
        ]
    mount root
    mapM_ (\k -> insertRecord items k (Item k)) ["a", "b", "c"]

File: guests/rust/reorder.rs
//! The reorder conformance scene (tools/scenes/reorder.steps).

#[derive(kaya::KayaGen, Clone, Debug, PartialEq)]
struct Item {
    title: String,
}

#[derive(Clone)]
enum Msg {
    Rotate,
    Lift,
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let items = ctx.apply(|tx| {
        let items = tx.collection::<Item>();
        // THE ROOT IS A ROW so the For's container is the only column:
        // construction order differs per language, column#0 may not.
        let root = tx.row(|tx| {
            let rotate = tx.button("rotate").id();
            msgs.on_click(rotate, Msg::Rotate);
            let lift = tx.button("lift").id();
            msgs.on_click(lift, Msg::Lift);
            for mut row in items.rows(tx) {
                row.label(Item::title());
            }
        })
        .id();
        tx.mount(root);
        for key in ["a", "b", "c"] {
            tx.insert(&items, key, Item { title: key.to_string() });
        }
        items
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Rotate => {
                ctx.apply(|tx| {
                    let entries = tx.items(&items);
                    let (first, _) =
                        entries.first().expect("reorder scene has entries").clone();
                    tx.move_to_end(&items, first);
                });
            }
            Msg::Lift => {
                ctx.apply(|tx| {
                    let entries = tx.items(&items);
                    let (last, _) =
                        entries.last().expect("reorder scene has entries").clone();
                    tx.move_to_front(&items, last);
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}
